### PR TITLE
Integrate nanopb to encode and decode metrics

### PR DESCRIPTION
**Pull request by anchitj**
### Changes done:

1. Added nanopb files in `src/nanopb` from https://github.com/nanopb/nanopb
2. Generated protobuf stubs using protoc compiler in nanopb and opentelemetry:1.0.0 protos, and added them in `src/opentelemetry`. One of the file metrics.proto needed an option present in metrics.options to create the callback needed for `oneof` types in .proto file which isn't generated by default.
3. Wrote encode related methods in `rdkafka_telemetry_encode.c`. Sample payload for a sum type metric will be like
```
metrics_data {
  resource_metrics {
    resource {
      attributes {
        key: "service.name"
        value {
          string_value: "example-service"
        }
      }
    }
    scope_metrics {
      scope {
        name: "example-scope"
        version: "1.0.0"
      }
      metrics {
        name: "requests_total"
        description: "Total number of requests"
        unit: "1"
        data {
          sum {
            data_points {
              attributes {
                key: "endpoint"
                value {
                  string_value: "/home"
                }
              }
              time_unix_nano: 1609462800123456789 
              value {
                as_int: 150 
              }
            }
            aggregation_temporality: AGGREGATION_TEMPORALITY_CUMULATIVE
          }
        }
      }
    }
  }
}
```
4. Wrote decode related methods in `rdkafka_telemetry_decode.c`. This is needed for testing, though not needed for functionality.
5. Only one metric is being encode right now by default.

____

### TODO:

Add prefix to both nanopb and opentelemetry structs to not cause conflicts with existing applications dependent on them.

### CODE REVIEW

File: src/rdkafka_telemetry_decode.c [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+ */
+
+#include "rd.h"
+#include "rdkafka_telemetry_decode.h"
+#include "pb.h"
+#include "pb_encode.h"
+#include "pb_decode.h"
+#include "metrics.pb.h"
+
+bool decode_and_print_string(pb_istream_t *stream,
```
Comment (milindl) [inline src/rdkafka_telemetry_decode.c:36]: Should we try to use rd_bool_t as much as possible? (for at least non-autogenerated code)

Also, I just realized that, if the nanopb library is using bool (internally _Bool), which was added in C99, we need to make sure it compiles with our older compilers (with -std=gnu90 and with Microsoft Visual Studio's compiler) :(

File: src/Makefile [hunk 2]
```diff
@@ -56,6 +59,9 @@ SRCS=		rdkafka.c rdkafka_broker.c rdkafka_msg.c rdkafka_topic.c \
 		rdvarint.c rdbuf.c rdmap.c rdunittest.c \
 		rdkafka_mock.c rdkafka_mock_handlers.c rdkafka_mock_cgrp.c \
 		rdkafka_error.c rdkafka_fetcher.c rdkafka_telemetry.c \
+		rdkafka_telemetry_encode.c rdkafka_telemetry_decode.c \
+		nanopb/pb_encode.c nanopb/pb_decode.c nanopb/pb_common.c \
+		opentelemetry/metrics.pb.c opentelemetry/common.pb.c opentelemetry/resource.pb.c \
```
[the remark this answers is not in the document]
Comment (milindl) [reply src/Makefile:64]: This will make the CI runs the windows compilation too

**Review by milindl — COMMENTED** [review]
Some comments, but will discuss this more with anchit

File: src/rdkafka_mock_handlers.c [hunk 3]
```diff
@@ -2230,7 +2244,7 @@ const struct rd_kafka_mock_api_handler
             {2, 2, -1, rd_kafka_mock_handle_OffsetForLeaderEpoch},
         [RD_KAFKAP_GetTelemetrySubscriptions] =
             {0, 0, 0, rd_kafka_mock_handle_GetTelemetrySubscriptions},
-        [RD_KAFKAP_PushTelemetry] = {0, 0, 0,
+        [RD_KAFKAP_PushTelemetry] = {0, 0, -1,
```
[the remark this answers is not in the document]
Comment (anchitj) [reply src/rdkafka_mock_handlers.c:2247]: Right, will fix this.

File: src/rdkafka_telemetry.c [hunk 4]
```diff
@@ -30,7 +30,9 @@
 #include "rdrand.h"
 #include "rdkafka_int.h"
 #include "rdkafka_telemetry.h"
+#include "rdkafka_telemetry_encode.h"
 #include "rdkafka_request.h"
+#include "pb.h"
```
[the remark this answers is not in the document]
Comment (anchitj) [reply src/rdkafka_telemetry.c:35]: Yeah makes sense, will update the import part to keep the directory as well.

File: src/rdkafka_telemetry_encode.c [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+        bool status = pb_get_encoded_size(
+            &message_size, opentelemetry_proto_metrics_v1_MetricsData_fields,
+            &metricsData);
+        if (!status) {
+                rd_kafka_dbg(rk, TELEMETRY, "METRICS",
+                             "Failed to get encoded size");
+                return NULL;
+        }
+
+        uint8_t *buffer = rd_malloc(message_size);
```
[the remark this answers is not in the document]
Comment (anchitj) [reply src/rdkafka_telemetry_encode.c:251]: Will update.

File: src/rdkafka_mock_handlers.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (anchitj) [reply src/rdkafka_mock_handlers.c]: Updated now.

File: src/rdkafka_telemetry_encode.c [hunk 6]
```diff
@@ -0,0 +1,274 @@
...
+
+        pb_ostream_t stream = pb_ostream_from_buffer(buffer, message_size);
+        status              = pb_encode(&stream,
+                           opentelemetry_proto_metrics_v1_MetricsData_fields,
+                           &metricsData);
+
+        if (!status) {
+                rd_kafka_dbg(rk, TELEMETRY, "METRICS", "Encoding failed: %s",
+                             PB_GET_ERROR(&stream));
+                free(buffer);
```
[the remark this answers is not in the document]
Comment (anchitj) [reply src/rdkafka_telemetry_encode.c]: Updated.

File: src/Makefile (hunk 2 above)
[the remark this answers is not in the document]
Comment (anchitj) [reply src/Makefile:64]: Updated windows build related files now. Seems like CI works for Windows.

File: src/rdkafka_telemetry_decode.c (hunk 1 above)
Comment (anchitj) [reply src/rdkafka_telemetry_decode.c:36]: Thanks, will use rd_bool_t in both encode and decode.

I compiled with `CFLAGS='-std=gnu90'` and tested with mock broker which is working.

File: src/rdkafka_telemetry.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (anchitj) [reply src/rdkafka_telemetry.c]: Updated

File: src/rdkafka_telemetry_encode.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (anchitj) [reply src/rdkafka_telemetry_encode.c]: Updated

File: src/opentelemetry/metrics.options [hunk 7]
```diff
@@ -0,0 +1 @@
+opentelemetry.proto.metrics.v1.Metric submsg_callback:true;
```
Comment (anchitj) [inline src/opentelemetry/metrics.options:1]: Add comment here

**Review by milindl — APPROVED** [review]
Looks good, mostly minor stuff but some larger changes also.

File: src/rdkafka_broker.h [hunk 8]
```diff
@@ -194,7 +194,8 @@ struct rd_kafka_broker_s { /* rd_kafka_broker_t */
         } rkb_c;
 
         struct {
-                /* TODO: fill this out. */
+                rd_atomic32_t connects; /**< Connection attempts,
```
[the remark this answers is not in the document]
Comment (anchitj) [reply src/rdkafka_broker.h]: Makes sense, update to int32_t.

File: src/rdkafka_request.c [hunk 9]
```diff
@@ -5261,16 +5261,19 @@ rd_kafka_PushTelemetryRequest(rd_kafka_broker_t *rkb,
         size_t len = sizeof(rd_kafka_uuid_t) + sizeof(int32_t) +
                      sizeof(rd_bool_t) + strlen(compression_type) +
                      metrics_size;
-        rkbuf = rd_kafka_buf_new_request(rkb, RD_KAFKAP_PushTelemetry, 1, len);
+        rkbuf = rd_kafka_buf_new_flexver_request(rkb, RD_KAFKAP_PushTelemetry,
```
[the remark this answers is not in the document]
Comment (anchitj) [reply src/rdkafka_request.c:5264]: Thanks, updated.

File: src/rdkafka_request.c [hunk 10]
```diff
@@ -5261,16 +5261,19 @@ rd_kafka_PushTelemetryRequest(rd_kafka_broker_t *rkb,
...
                                strlen(compression_type));
-        rd_kafka_buf_write_bytes(rkbuf, metrics ? metrics : "", metrics_size);
 
-        rd_kafka_buf_ApiVersion_set(rkbuf, ApiVersion, 0);
+        rd_kafkap_bytes_t *metric_bytes =
+            rd_kafkap_bytes_new(metrics, metrics_size);
+        rd_kafka_buf_write_kbytes(rkbuf, metric_bytes);
+        rd_free(metric_bytes);
 
         /* Let the handler perform retries so that it can pick
```
[the remark this answers is not in the document]
Comment (anchitj) [reply src/rdkafka_request.c]: Removed the comment.

File: src/rdkafka_telemetry.c [hunk 11]
```diff
@@ -192,17 +194,20 @@ static void rd_kafka_send_push_telemetry(rd_kafka_t *rk,
         // serialize.
 
         // TODO: Update dummy values
-        void *metrics_payload  = NULL;
+        size_t metrics_payload_size;
+        void *metrics_payload  = encode_metrics(rk, &metrics_payload_size);
         char *compression_type = "gzip";
```
[the remark this answers is not in the document]
Comment (anchitj) [reply src/rdkafka_telemetry.c]: Right. I've added a TODO for this.

File: src/rdkafka_telemetry.c [hunk 12]
```diff
@@ -192,17 +194,20 @@ static void rd_kafka_send_push_telemetry(rd_kafka_t *rk,
         // serialize.
 
         // TODO: Update dummy values
-        void *metrics_payload  = NULL;
+        size_t metrics_payload_size;
+        void *metrics_payload  = encode_metrics(rk, &metrics_payload_size);
```
[the remark this answers is not in the document]
Comment (anchitj) [reply src/rdkafka_telemetry.c]: Thanks, updated.

File: src/rdkafka_telemetry_decode.h [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef _RDKAFKA_RDKAFKA_TELEMETRY_DECODE_H
+#define _RDKAFKA_RDKAFKA_TELEMETRY_DECODE_H
+
+void decode_metric(void *buffer, size_t size);
```
[the remark this answers is not in the document]
Comment (anchitj) [reply src/rdkafka_telemetry_decode.h]: Renamed.

File: src/rdkafka_telemetry_encode.c [hunk 14]
```diff
@@ -0,0 +1,294 @@
...
+
+#include "rd.h"
+#include "rdkafka_int.h"
+#include "rdkafka_telemetry_encode.h"
+#include "nanopb/pb.h"
+#include "nanopb/pb_encode.h"
+#include "nanopb/pb_decode.h"
+#include "opentelemetry/metrics.pb.h"
+
+rd_bool_t
```
[the remark this answers is not in the document]
Comment (anchitj) [reply src/rdkafka_telemetry_encode.c]: Makes sense, missed doing this. Made all the file level functions static.

File: src/rdkafka_telemetry_encode.c [hunk 15]
```diff
@@ -0,0 +1,294 @@
...
+        return pb_encode_string(stream, (uint8_t *)(*arg), strlen(*arg));
+}
+
+// TODO: Update to handle multiple data points.
+rd_bool_t encode_number_data_point(pb_ostream_t *stream,
+                                   const pb_field_t *field,
+                                   void *const *arg) {
+        opentelemetry_proto_metrics_v1_NumberDataPoint *data_point =
+            (opentelemetry_proto_metrics_v1_NumberDataPoint *)*arg;
+        if (!pb_encode_tag_for_field(stream, field)) {
```
[the remark this answers is not in the document]
Comment (anchitj) [reply src/rdkafka_telemetry_encode.c]: Updated to this convention everywhere.

File: src/rdkafka_telemetry_encode.c [hunk 16]
```diff
@@ -0,0 +1,294 @@
...
+                return false;
+        }
+        opentelemetry_proto_common_v1_KeyValue *key_value =
+            (opentelemetry_proto_common_v1_KeyValue *)*arg;
+        return pb_encode_submessage(
+            stream, opentelemetry_proto_common_v1_KeyValue_fields, key_value);
+}
+
+static const char *rd_kafka_type2str(rd_kafka_type_t type) {
```
[the remark this answers is not in the document]
Comment (anchitj) [reply src/rdkafka_telemetry_encode.c]: Moved to rdkafka_int.h

File: src/rdkafka_telemetry_encode.c [hunk 17]
```diff
@@ -0,0 +1,294 @@
...
+        size_t message_size;
+        uint8_t *buffer;
+        pb_ostream_t stream;
+        rd_bool_t status;
+        // TODO: Removed hardcoded values
+        char *metric_suffix = ".connection.creation.total",
+             *metric_description =
+                 "The total number of connections established.",
+             *metric_unit     = "1",
+             *metric_type_str = rd_kafka_type2str(rk->rk_type), *metric_name;
```
[the remark this answers is not in the document]
Comment (anchitj) [reply src/rdkafka_telemetry_encode.c]: Thanks, updated.

**Review by milindl — APPROVED** [review]

**Pull request merged**